src/storage/sql_table.cpp

```diff
@@ -131,109 +140,219 @@ TupleSlot SqlTable::Insert(const common::ManagedPointer<transaction::Transaction
+                       parser::ExpressionType::VALUE_CONSTANT,
+                   " For now, we only handle constant default values.");
+
+    // Find the until a version has the required column, use the default value that is closest to the tuple version
+    for (layout_version_t start_version = tuple_version + 1; start_version <= layout_version; start_version++) {
+      auto curr_version = tables_.at(start_version);
+      // Not found in this version
+      if (curr_version.default_value_map_.find(it.second) == curr_version.default_value_map_.end()) continue;
+
+      // Found the default value at curr_version
+      auto default_const = curr_version.default_value_map_.at(it.second)
+                               .CastManagedPointerTo<const parser::ConstantValueExpression>()
+                               ->GetValue();
+      auto value_size = curr_version.schema_->GetColumn(it.second).AttrSize();
+
+      // zero out the temporary buffer before peeking
+      byte output[value_size];
+      memset(output, 0, value_size);
+
+      if (PeekValue(default_const, &(output[0]))) {
+        StorageUtil::CopyWithNullCheck(output, out_buffer, curr_version.schema_->GetColumn(it.second).AttrSize(),
+                                       it.first);
+        out_buffer->SetNotNull(it.first);
+        break;
+      }
     }
   }
 }
 
-template void SqlTable::FillMissingColumns<ProjectedRow>(ProjectedRow *out_buffer,
-                                                         const DataTableVersion &desired_version) const;
-template void SqlTable::FillMissingColumns<ProjectedColumns::RowView>(ProjectedColumns::RowView *out_buffer,
-                                                                      const DataTableVersion &desired_version) const;
+template void SqlTable::FillMissingColumns<ProjectedRow>(
+    ProjectedRow *out_buffer, const std::vector<std::pair<size_t, catalog::col_oid_t>> &missingl_cols,
+    layout_version_t tuple_version, layout_version_t layout_version) const;
+template void SqlTable::FillMissingColumns<ProjectedColumns::RowView>(
+    ProjectedColumns::RowView *out_buffer, const std::vector<std::pair<size_t, catalog::col_oid_t>> &missingl_cols,
+    layout_version_t tuple_version, layout_version_t layout_version) const;
 
 template <class RowType>
-bool SqlTable::AlignHeaderToVersion(RowType *const out_buffer, const DataTableVersion &tuple_version,
-                                    const DataTableVersion &desired_version, col_id_t *cached_ori_header) const {
-  bool missing_col = false;
-  // reserve the original header, aka intended column ids'
-  std::memcpy(cached_ori_header, out_buffer->ColumnIds(), sizeof(col_id_t) * out_buffer->NumColumns());
-
-  // for each column id in the intended version of datatable, change it to match the current schema version
+std::vector<std::pair<size_t, catalog::col_oid_t>> SqlTable::AlignHeaderToVersion(
```

Comment:
Consider returning a std::vector<size_t> and then getting the col_oid_t later as needed in this class

Reply:
I don't see why this is necessary. The current design works fine